src/hashset.h

```diff
@@ -0,0 +1,205 @@
+ *         function and how to get the key in an element.
+ */
+
+#include "fmacros.h"
+#include <stddef.h>
+#include <stdint.h>
+
+/* --- Opaque types --- */
+
+typedef struct hashset hashset;
+
+/* --- Non-opaque types --- */
+
+/* The hashsetType is a set of callbacks for a hashset. All callbacks are
+ * optional. With all callbacks omitted, the hashset is effectively a set of
+ * pointer-sized integers. */
+typedef struct {
+    /* If the type of an element is not the same as the type of a key used for
+     * lookup, this callback needs to return the key within an element. */
+    const void *(*elementGetKey)(const void *element);
+    /* Hash function. Defaults to hashing the bits in the pointer, effectively
+     * treating the pointer as an integer. */
+    uint64_t (*hashFunction)(const void *key);
+    /* Compare function, returns 0 if the keys are equal. Defaults to just
+     * comparing the pointers for equality. */
+    int (*keyCompare)(hashset *t, const void *key1, const void *key2);
+    /* Callback to free an element when it's overwritten or deleted.
+     * Optional. */
+    void (*elementDestructor)(hashset *t, void *elem);
+    /* Optional callback to control when resizing should be allowed. */
+    int (*resizeAllowed)(size_t moreMem, double usedRatio);
+    /* Invoked at the start of rehashing. Both tables are already created. */
+    void (*rehashingStarted)(hashset *t);
+    /* Invoked at the end of rehashing. Both tables still exist and are cleaned
+     * up after this callback. */
+    void (*rehashingCompleted)(hashset *t);
+    /* Allow a hashset to carry extra caller-defined metadata. The extra memory
+     * is initialized to 0. */
+    size_t (*getMetadataSize)(void);
+    /* Flag to disable incremental rehashing */
+    unsigned instant_rehashing : 1;
+    /* Allow the caller to store some data here in the type. It's useful for the
+     * rehashingStarted and rehashingCompleted callbacks. */
+    void *userdata;
+} hashsetType;
+
+typedef enum {
+    HASHSET_RESIZE_ALLOW = 0,
+    HASHSET_RESIZE_AVOID,
+    HASHSET_RESIZE_FORBID,
+} hashsetResizePolicy;
+
+typedef void (*hashsetScanFunction)(void *privdata, void *element);
+
+/* Constants */
+#define HASHSET_BUCKET_SIZE 64 /* bytes */
+
+/* Scan flags */
+#define HASHSET_SCAN_EMIT_REF (1 << 0)
+#define HASHSET_SCAN_SINGLE_STEP (1 << 2)
+
+typedef struct {
+    hashset *hashset;
+    long index;
+    int table;
+    int posInBucket;
+    /* unsafe iterator fingerprint for misuse detection. */
+    uint64_t fingerprint;
+    int safe;
+} hashsetIterator;
+
+typedef struct hashsetStats {
+    int htidx;
+    unsigned long buckets;       /* num buckets */
+    unsigned long maxChainLen;   /* probing chain length */
+    unsigned long totalChainLen; /* buckets with probing flag */
+    unsigned long htSize;        /* buckets * positions-per-bucket */
+    unsigned long htUsed;        /* num elements */
+    unsigned long *clvector;
```

Comment:
```suggestion
    unsigned long *clvector; /* Whatever a clvector is */
```
I think it's funny the least clear one is the one without documentation :D

Reply:
Yes.... this stuff wasn't the most interesting part of the implementation. It's very much like the dict and it's used by DEBUG HTSTATS.

clvector = chain length vector, i.e. each index i in the vector records the number of chains of length i. For dict, this was bucket chain length, but here I used it for probe chain length. This way, we can present something similar in DEBUG HTSTATS.